### PR TITLE
Create user on identity check answers form submission

### PR DESCRIPTION
**Pull request by wilpow-github**
### Context

As a user
I want to create an Identity account
So that I can sign in to Teacher Services services

### Changes proposed in this pull request

When a user clicks Continue on the Check Answers page, create a User record and save it into the Identity database.

### Checklist

-   [x] Attach to Trello card
-   [ ] Rebased master
-   [ ] Cleaned commit history
-   [ ] Tested by running locally
-   [ ] Reminder created to manually clean any removed app settings post deployment

### CODE REVIEW

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs [hunk 1]
```diff
@@ -31,8 +43,48 @@ public void OnGet()
...
+            User = user
+        });
+
+        try
+        {
+            await _dbContext.SaveChangesAsync();
+        }
+        catch (DbUpdateException dex) when (dex.IsUniqueIndexViolation("ix_users_email_address"))
+        {
+            return Redirect(_linkGenerator.TrnInUseChooseEmail());
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs:84]: Not doing any exception handling here except redirecting to TrnInUseChooseEmail

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Trn/CheckAnswersTests.cs [hunk 2]
```diff
@@ -291,4 +293,63 @@ public async Task Get_ValidRequestWithAwardedQtsFalse_DoesNotShowProviderRow()
...
+
+        // Assert
+        Assert.Equal(StatusCodes.Status302Found, (int)response.StatusCode);
+        Assert.Equal(authStateHelper.GetNextHopUrl(), response.Headers.Location?.OriginalString);
+
+        await TestData.WithDbContext(async dbContext =>
+        {
+            var user = await dbContext.Users.Where(u => u.EmailAddress == authStateHelper.AuthenticationState.EmailAddress).SingleOrDefaultAsync();
+            Assert.NotNull(user);
+            Assert.Equal(authStateHelper.AuthenticationState.OfficialFirstName, user.FirstName);
```
Comment (wilpow-github) [inline dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Trn/CheckAnswersTests.cs:321]: Not really validating the value of any other columns

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs [hunk 3]
```diff
@@ -31,8 +43,48 @@ public void OnGet()
...
+            FirstName = string.IsNullOrEmpty(authenticationState.FirstName) ? authenticationState.OfficialFirstName! : authenticationState.FirstName,
+            LastName = string.IsNullOrEmpty(authenticationState.LastName) ? authenticationState.OfficialLastName! : authenticationState.LastName,
+            Updated = _clock.UtcNow,
+            UserId = userId,
+            UserType = UserType.Default,
+            Trn = authenticationState.StatedTrn,
+            TrnAssociationSource = !string.IsNullOrEmpty(authenticationState.Trn) ? TrnAssociationSource.Lookup : null,
+            LastSignedIn = _clock.UtcNow,
+            RegisteredWithClientId = authenticationState.OAuthState?.ClientId,
+            TrnLookupStatus = !string.IsNullOrEmpty(authenticationState.Trn) ? TrnLookupStatus.Found : TrnLookupStatus.None
```
Comment (wilpow-github) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs:63]: This is slightly changed from TrnCallback as we don't have SupportTicketCreated

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs (hunk 1 above)
Comment (gunndabad) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs:81]: Need to replicate what's happening in `TrnCallback` here (find the user with the existing email, generate PIN etc.)

File: dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs [hunk 4]
```diff
@@ -31,8 +40,48 @@ public void OnGet()
...
+            CompletedTrnLookup = _clock.UtcNow,
+            Created = _clock.UtcNow,
+            DateOfBirth = authenticationState.DateOfBirth,
+            EmailAddress = authenticationState.EmailAddress!,
+            FirstName = string.IsNullOrEmpty(authenticationState.FirstName) ? authenticationState.OfficialFirstName! : authenticationState.FirstName,
+            LastName = string.IsNullOrEmpty(authenticationState.LastName) ? authenticationState.OfficialLastName! : authenticationState.LastName,
+            Updated = _clock.UtcNow,
+            UserId = userId,
+            UserType = UserType.Default,
+            Trn = authenticationState.StatedTrn,
```
Comment (gunndabad) [inline dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs:59]:
```suggestion
            Trn = authenticationState.Trn,
```

Comment (wilpow-github) [reply dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Trn/CheckAnswers.cshtml.cs:59]: Ah yes forgot to change after rebasing onto main

File: dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Trn/CheckAnswersTests.cs [hunk 5]
```diff
@@ -291,4 +293,60 @@ public async Task Get_ValidRequestWithAwardedQtsFalse_DoesNotShowProviderRow()
...
+        {
+            Content = new FormUrlEncodedContentBuilder()
+        };
+
+        // Act
+        var response = await HttpClient.SendAsync(request);
+
+        // Assert
+        Assert.Equal(StatusCodes.Status302Found, (int)response.StatusCode);
+        Assert.StartsWith("/sign-in/trn/different-email", response.Headers.Location?.OriginalString);
```
Comment (wilpow-github) [inline dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Trn/CheckAnswersTests.cs:344]: Ticket mentions choose-email page, but I believe this is correct

Comment (gunndabad) [reply dotnet-authserver/tests/TeacherIdentity.AuthServer.Tests/EndpointTests/SignIn/Trn/CheckAnswersTests.cs:344]: I got the ticket wrong; should be `/sign-in/trn/choose-email` I think

**Review by gunndabad — APPROVED** [review]

**Pull request merged**